### DISCUSSION TITLE
How can I precisely measure the performance of instructions for data transmission between CPU and GPU, like queue.writeBuffer()?

### DISCUSSION THREAD

**Discussion by csAugust** [gpuweb/gpuweb#4709] [Q&A]
I wish to know how much time data transmission between CPU and GPU costs so that I can try to improve the overall WebGPU programs' performance. But till now I only know few ways to measure it and I don't really know are they correct. 

- Use "timestamp-query" feature. It seems that it can only measure compute passes' cost as `GPUComputePassTimestampWrites` only has 2 time point for measurement (pass beginning and end). So it can't be used for instructions like `queue.writeBuffer()`. See also https://stackoverflow.com/questions/76713260/how-to-time-profile-see-when-tasks-are-complete-in-wgpu-targeting-web.
- Use `queue.onSubmittedWorkDone()`before and after the `queue.writeBuffer()` and calculate the time cost in JS promise callbacks. I don't know is it correct and it seems not precise intuitively.
- Use other debugger tools, like DevTools when running WebGPU programs in Chrome. In DevTools I can only get a very sketchy view about how much time it costs before and after GPU executing some tasks.

So I wonder to know how to measure the time cost of:

- Transmitting data from CPU to GPU, including `createBuffer()` and `queue.writeBuffer()`;
- Transmitting data from GPU to CPU, including `createBuffer()` for staging buffer and `buffer.mapAsync()` when reading.

I've tried searched keywords like 'track''performace''writeBuffer' but can't get an answer. If any one can give me some ideas or figure out any mistakes in my thought I'll be much appreciated. Thank you!

**Comment by greggman** [comment]
I don't know how to measure that but unrelated, you might find [this article](https://webgpufundamentals.org/webgpu/lessons/webgpu-optimization.html) useful.

As for reading, there's also [this thread](https://github.com/gpuweb/gpuweb/discussions/1428#discussioncomment-8977501)

**Reply by greggman** [reply]
Ah, should add, there is https://perfetto.dev/

Turn on at least

* Javascript execution
* Web content rendering, layout and compositing
* UI rendering & surface compositing

This info is only used in Chromium based browsers.

Not sure whether or not that's helpful as there's too much data

**Reply by csAugust** [reply]
Thank you for your reply!  The first two links offer some useful optimization techniques which I can try later. I haven't used perfetto, and it seems a little complex. I'm trying to get used to it.

**Comment by mmgeorge** [comment]
Running into the same issue, doesn't seem like we have a way to time events submitted directly onto the queue? In particular I was looking to time [queue.copyExternalImageToTexture ](https://www.w3.org/TR/webgpu/#dom-gpuqueue-copyexternalimagetotexture). Wondering if we could wrap in an [error scope](https://www.w3.org/TR/webgpu/#error-scopes) and time that...?

Or maybe we can wrap in empty compute / render pipelines with just timers?

**Reply by mmgeorge** [reply]
Maybe we need to wait on https://github.com/gpuweb/gpuweb/issues/3952

**Comment by Frendysemon** [comment]
Yeah writeBuffer is hard to time directly. onSubmittedWorkDone is useful for a rough estimate but not exact transfer time.
